File: test/language/choice_types/python/FunctionReturningLiteralSelectorChoiceTest.py
import ChoiceTypes

class FunctionReturningLiteralSelectorChoiceTest(ChoiceTypes.TestCase):
    def testField8(self):
        selector = self.api.Selector(False)
        testChoice = self.api.TestChoice(selector)
        testChoice.field8 = 0x7F
        self.assertEqual(0x7F, testChoice.field8)
        self.assertEqual(8, testChoice.bitsizeof())

    def testField16(self):
        selector = self.api.Selector(True)
        testChoice = self.api.TestChoice(selector, field16_=0x7F7F)
        self.assertEqual(0x7F7F, testChoice.field16)
        self.assertEqual(16, testChoice.bitsizeof())

File: test/language/choice_types/cpp/FunctionReturningLiteralSelectorChoiceTest.cpp
#include "gtest/gtest.h"

#include "choice_types/function_returning_literal_selector_choice/TestChoice.h"

// just testChoice setters and getters
namespace choice_types
{
namespace function_returning_literal_selector_choice
{

class FunctionReturningLiteralSelectorChoiceTest : public ::testing::Test
{
protected:
    struct SelectorHolder
    {
        SelectorHolder(Selector::ParameterExpressions selectorParameterExpressions)
        {
            selector.initialize(selectorParameterExpressions);
            choiceParameterExpressions ={ this, 0, getSelector };
        }

        static Selector& getSelector(void* owner, size_t)
        {
            return static_cast<SelectorHolder*>(owner)->selector;
        }

        Selector selector;
        TestChoice::ParameterExpressions choiceParameterExpressions;
    };
};

TEST_F(FunctionReturningLiteralSelectorChoiceTest, field8)
{
    SelectorHolder selectorHolder({ nullptr, 0, [](void*, size_t) { return false; } });
    TestChoice testChoice;
    testChoice.initialize(selectorHolder.choiceParameterExpressions);
    testChoice.setField8(0x7F);
    ASSERT_EQ(0x7F, testChoice.getField8());
    ASSERT_EQ(8, testChoice.bitSizeOf());
}

TEST_F(FunctionReturningLiteralSelectorChoiceTest, field16)
{
    SelectorHolder selectorHolder({ nullptr, 0, [](void*, size_t) { return true; } });
    TestChoice testChoice;
    testChoice.initialize(selectorHolder.choiceParameterExpressions);
    testChoice.setField16(0x7F7F);
    ASSERT_EQ(0x7F7F, testChoice.getField16());
    ASSERT_EQ(16, testChoice.bitSizeOf());
}

} // namespace function_returning_literal_selector_choice
} // namespace choice_types
